File: Task1C/task1c/src/main/java/com/cps2004/Food.java
package com.cps2004;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

public abstract class Food extends Product {
    private double weight;
    private String expirationDate;

    // Default constructor
    public Food() {
        super();
        this.weight = 0;
        this.expirationDate = "";
    }

    // Parameterized constructor
    public Food(String name, int id, Supplier supplier, int price, int volume, double weight, String expirationDate) {
        super(name, id, supplier, price, volume);
        this.weight = weight;
        this.expirationDate = expirationDate;
    }

    // Getters and setters
    public double getWeight() {
        return weight;
    }

    public String getExpirationDate() {
        return expirationDate;
    }

    public void setWeight(double weight) {
        this.weight = weight;
    }

    public void setExpirationDate(String expirationDate) {
        this.expirationDate = expirationDate;
    }

    @Override
    // Calculate discounted price
    public abstract double calculate_discount(String month, int quantity);

    // Display product information
    public void display_product_info() {
        super.display_product_info();
        System.out.println("Product weight: " + weight);
        System.out.println("Product expiration date: " + expirationDate);
    }
    // Save product information to file
    public void save(ObjectNode newItem) {
        super.save(newItem);
        newItem.put("weight", this.weight);
        newItem.put("expirationDate", this.expirationDate);
    }

    // Load product information from file
    public void load(JsonNode node) {
        super.load(node);
        weight = node.get("weight").asDouble();
        expirationDate = node.get("expirationDate").asText();
    }
}

File: Task3C/food.cpp
#include "food.h"

// Default constructor
Food::Food() {
    this->weight = 0;
    this->expirationDate = "";
}

// Parameterised constructor
Food::Food(string name, int price, Supplier* supplier, int volume, int productID, double weight, string expirationDate) : Product(name, price, supplier, volume, productID) {
    this->weight = weight;
    this->expirationDate = expirationDate;
}

// Getters and setters
double Food::getWeight() {
    return this->weight;
}

string Food::getExpirationDate() {
    return this->expirationDate;
}

void Food::setWeight(double weight) {
    this->weight = weight;
}

void Food::setExpirationDate(string expirationDate) {
    this->expirationDate = expirationDate;
}

// Display product information
void Food::display_product_info() {
    Product::display_product_info();
    cout << "Weight: " << this->weight << endl;
    cout << "Expiration Date: " << this->expirationDate << endl;
}

// Save product information to file
void Food::save(ostream& out) const {
    Product::save(out);
    out << weight << "," << expirationDate << "\n";
}

// Load product information from file
void Food::load(istream& in) {
    Product::load(in);

    string line;
    getline(in, line, '\n');
    stringstream ss(line);

    string token;
    getline(ss, token, ',');
    weight = stod(token);
    
    getline(ss, token, ',');
    expirationDate = token;
}